lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,268 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module PowerShell
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @stdin, @stdout, @stderr, @ps_process = Open3.popen3(cmd)
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def make_ps_code(powershell_code, output_ready_event_name, err_ready_event_name, timeout_ms = 300 * 1000)
+        # always need a trailing newline to ensure PowerShell parses code
+        code = <<-CODE
+        $outputEvent = [Threading.EventWaitHandle]::OpenExisting("#{output_ready_event_name}")
+        $errEvent = [Threading.EventWaitHandle]::OpenExisting("#{err_ready_event_name}")
+        if ($runspace -eq $null)
+        {
+          # CreateDefault2 requires PS3
+          if ([InitialSessionState].GetMethod('CreateDefault2'))
+          {
+            $sessionState = [InitialSessionState]::CreateDefault2()
+          }
+          else
+          {
+            $sessionState = [InitialSessionState]::CreateDefault()
+          }
+          $runspace = [RunspaceFactory]::CreateRunspace($sessionState)
+          $runspace.Open()
+        }
+
+        $powershell_code = @'
+#{powershell_code}
+'@
+        try
+        {
+          $ps = $null
+
+          $file = [IO.Path]::GetTempFileName()
+          $file = $file -replace ".tmp",".ps1"
+          $fileStream = [IO.File]::Open($file, [IO.FileMode]::OpenOrCreate, [IO.FileAccess]::Write, [IO.FileShare]::Read)
+          $sw = New-Object IO.StreamWriter ($fileStream, [Text.Encoding]::UTF8)
+          $sw.Write($powershell_code)
+          $sw.Flush()
+          $sw.Close()
+          $sw.Dispose()
```

Comment:
I would have suggested that earlier but I didn't know what kind of undertaking you wanted to do here.